### PR TITLE
flow : use xxHash and eBPF fixies

### PR DESCRIPTION
**Pull request by andrekassis**
Hi guys,
as promised this commit includes a few changes. 
Flow table is now bounded - no memory leaks
Hash function was replaced - no collisions 
Only one copy of the flow table is kept in ebpf now

This is still the first commit. Still need to report the number of deleted flows so this is on our todo list.
Cheers

### CODE REVIEW

**Comment by andrekassis** [comment]
Oh, I added xxhash vi go get. need to also add this to the makefile
will do tomorrow

File: flow/flow.go [hunk 1]
```diff
@@ -309,28 +310,43 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
 	return layer.TransportFlow(), nil
 }
 
+//Hash flow with custom func
+func Hash(f gopacket.Flow, hasher hash.Hash) {
+	if strings.Compare(string(f.Src().Raw()), string(f.Dst().Raw())) > 0 {
```
Comment (nplanel) [inline flow/flow.go]: may retrieve the the endpoints only once with src, dst := f.Endpoints()

File: flow/flow.go [hunk 2]
```diff
@@ -309,28 +310,43 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
 	return layer.TransportFlow(), nil
 }
 
+//Hash flow with custom func
+func Hash(f gopacket.Flow, hasher hash.Hash) {
```
Comment (nplanel) [inline flow/flow.go]: on side note, you should propose a PR to gopacket to use xxHash  : github.com/google/gopacket/flows.go

File: flow/flow.go [hunk 3]
```diff
@@ -309,28 +310,43 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
...
 func (p *Packet) Key(parentUUID string, opts Opts) string {
-	var uuid uint64
-
+	hasher := xxHash64.New(0)
 	// uses L2 is requested or if there is no network layer
 	if opts.LayerKeyMode == L2KeyMode || p.NetworkLayer() == nil {
 		if layer := p.LinkLayer(); layer != nil {
-			uuid ^= layer.LinkFlow().FastHash()
+			Hash(layer.LinkFlow(), hasher)
+			hasher.Write([]byte(strconv.FormatInt(linkID(p), 10)))
```
Comment (nplanel) [inline flow/flow.go]: use binary package to convert int64 to bytes
     b := make([]byte, 8)
    binary.BigEndian.PutUint64(b, linkID(p))

File: flow/flow.go [hunk 4]
```diff
@@ -309,28 +310,43 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
...
+	if layer := p.TransportLayer(); layer != nil {
+		Hash(layer.TransportFlow(), hasher)
 	}
 	if af, err := p.ApplicationFlow(); err == nil {
-		uuid ^= af.FastHash()
+		Hash(af, hasher)
 	}
-
-	return parentUUID + strconv.FormatUint(uuid, 10)
+	return hex.EncodeToString(hasher.Sum(nil))
```
Comment (nplanel) [inline flow/flow.go]: you should keep parentUUID logic or you will break tunnel

File: flow/probes/ebpf.go [hunk 5]
```diff
@@ -394,50 +109,28 @@ func (p *EBPFProbe) run() {
...
 				}
-			}
-
-			for k, v := range ebpfFlows {
-				if time.Now().Sub(v.last).Seconds() > p.expire.Seconds() {
-					delete(ebpfFlows, k)
+				flow.SetEBPFFlow(&ebpfFlows[i], us, last, unsafe.Pointer(&kernFlows[i]), startKTimeNs, p.probeNodeTID)
+				flowChanOperation <- &ebpfFlows[i]
+				nextAvailablePtr++
+				if nextAvailablePtr == 2001 {
```
Comment (nplanel) [inline flow/probes/ebpf.go]: please set a constant like FLOW_POOL_SIZE
nextAvailablePtr is not used, should be used line 400
nextAvailablePtr = nextAvailablePtr++ % FLOW_POOL_SIZE

File: flow/table.go [hunk 6]
```diff
@@ -113,29 +120,280 @@ func updateTCPFlagTime(prevFlagTime int64, currFlagTime int64) int64 {
 	return currFlagTime
 }
 
+func kernFlowKeyOuter(kernFlow *C.struct_flow) string {
```
Comment (nplanel) [inline flow/table.go]: please keep eBPF probe code in the probe itself

File: flow/probes/ebpf.go [hunk 7]
```diff
@@ -394,50 +109,28 @@ func (p *EBPFProbe) run() {
...
+				if us.After(now) {
+					us = now
 				}
-			}
-
-			for k, v := range ebpfFlows {
-				if time.Now().Sub(v.last).Seconds() > p.expire.Seconds() {
-					delete(ebpfFlows, k)
+				flow.SetEBPFFlow(&ebpfFlows[i], us, last, unsafe.Pointer(&kernFlows[i]), startKTimeNs, p.probeNodeTID)
+				flowChanOperation <- &ebpfFlows[i]
```
Comment (nplanel) [inline flow/probes/ebpf.go]: not sure the allocation pool will really enhance something here or it need to match the chan size

File: flow/probes/ebpf.go (hunk 5 above)
Comment (nplanel) [reply flow/probes/ebpf.go]: you should test you got an potential overlap ( (nextAvailablePtr+1) % FLOW_POOL_SIZE == i)
meaning your pool is full

**Comment by nplanel** [comment]
@andrekassis could you re-run your tests with latest fix PR #1803

File: flow/flow.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrekassis) [reply flow/flow.go]: done

Comment (andrekassis) [reply flow/flow.go]: done

File: flow/flow.go (hunk 2 above)
Comment (andrekassis) [reply flow/flow.go]: will do, thanks;)

File: flow/flow.go (hunk 3 above)
Comment (andrekassis) [reply flow/flow.go]: done

File: flow/flow.go (hunk 4 above)
Comment (andrekassis) [reply flow/flow.go]: done

File: flow/probes/ebpf.go (hunk 5 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: done for the first issue.
As for the second, there shouldn't be any overlaps since the channel is FIFO and blocks when it's full (after 1000) the 2001 is merely for safety. Am I correct?

File: flow/table.go (hunk 6 above)
Comment (andrekassis) [reply flow/table.go]: EBPFFlow is now part of the table isn't it?

File: flow/probes/ebpf.go (hunk 7 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: extra size is for safety. Althouth the channel is guaranteed to be fifi, when using a size matching that of the channel (i.e. 1000) there was indeed an overlap. maybe because reads happen in bulks, so that when the channel is full it is drained on a lockstep (perhaps for performance gain). so using 2001 prevents that from happening. (tested)

File: gremlin/traversal/cflows.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"math/rand"
+	"net"
+	"time"
+	"unsafe"
+
+	"github.com/skydive-project/skydive/flow"
+)
+
+func newICMPFlow(id uint32, nodeTID string, linkSrc string, linkDst string) *flow.EBPFFlow {
```
Comment (safchain) [inline gremlin/traversal/cflows.go]: can you move this in the test file as it is only used for the tests

File: flow/table.go [hunk 9]
```diff
@@ -268,22 +531,24 @@ func (ft *Table) update(updateFrom, updateTime int64) {
 	logging.GetLogger().Debugf("flow table update: %d, %d", updateFrom, updateTime)
 
 	var updatedFlows []*Flow
-	for k, f := range ft.table {
+	for _, k := range ft.table.Keys() {
```
Comment (safchain) [inline flow/table.go]: This will create a copy of all the keys. It would be nice to have an implementation allowing to iterate over the values/keys

File: flow/table.go [hunk 10]
```diff
@@ -113,29 +120,280 @@ func updateTCPFlagTime(prevFlagTime int64, currFlagTime int64) int64 {
...
+			layersPath += "GRE"
+		default:
+			layersPath += "Unknown"
+		}
+	}
+	return layersPath, hasGRE
+}
+
+func (ft *Table) newFlowFromEBPF(ebpfFlow *EBPFFlow, key string) *Flow {
+
```
Comment (safchain) [inline flow/table.go]: nit: remove extra line

File: flow/table.go [hunk 11]
```diff
@@ -113,29 +120,280 @@ func updateTCPFlagTime(prevFlagTime int64, currFlagTime int64) int64 {
...
+				parent.Network = &FlowLayer{
+					Protocol: FlowProtocol_IPV6,
+					A:        net.IP(netA).String(),
+					B:        net.IP(netB).String(),
+				}
+			}
+		}
+		pkey := kernFlowKeyOuter(ebpfFlow.kernFlow)
+		parent.UpdateUUID(pkey, Opts{LayerKeyMode: L3PreferedKeyMode})
+		ft.table.Add(pkey, parent)
```
Comment (safchain) [inline flow/table.go]: I would prefer to have either both flows Added in this function or both returned, but here the parent is inserted here and the other flow in the caller

File: flow/probes/ebpf.go [hunk 12]
```diff
@@ -368,10 +81,13 @@ func (p *EBPFProbe) run() {
 	var startKTimeNs int64
 	var start time.Time
 
-	ebpfFlows := make(map[C.__u64]*EBPFFlow)
 	updateTicker := time.NewTicker(ebpfUpdate)
 	defer updateTicker.Stop()
 
+	kernFlows := [2001]C.struct_flow{}
```
Comment (safchain) [inline flow/probes/ebpf.go]: this size should be related to the chan size. I'm afraid that one day we will chan the chan size of this one without changing the other

File: flow/table.go [hunk 13]
```diff
@@ -43,6 +52,21 @@ type Handler struct {
 	callback ExpireUpdateFunc
 	every    time.Duration
 }
+type EBPFFlow struct {
```
Comment (safchain) [inline flow/table.go]: can you move all the ebpf related functions into a dedicated file, it will be easier to read

File: gremlin/traversal/cflows.go (hunk 8 above)
Comment (andrekassis) [reply gremlin/traversal/cflows.go]: The reason this funtion is in a seperate file is that cgo is not allowed in test files. i.e. files ending with _test.go
so this is my work around for this. If you have a better idea would love to hear it;)

**Comment by andrekassis** [comment]
run 
skydive-compile-tests

File: flow/table.go [hunk 14]
```diff
@@ -493,8 +755,8 @@ func (ft *Table) Run() {
 			}
 		case ps := <-ft.packetSeqChan:
 			ft.processPacketSeq(ps)
-		case op := <-ft.flowChanOperation:
-			ft.processFlowOP(op)
+		case ebpfFlow := <-ft.flowEBPFChan:
```
Comment (safchain) [inline flow/table.go]: please do not remove the flowOpChan as it will be used for other probes

File: gremlin/traversal/cflows.go (hunk 8 above)
Comment (safchain) [reply gremlin/traversal/cflows.go]: That is indeed a good reason :)

File: gremlin/traversal/flow_test.go [hunk 15]
```diff
@@ -132,8 +122,8 @@ func TestHasStep(t *testing.T) {
 		time.Sleep(100 * time.Millisecond)
 	}
 
-	flowChan <- &flow.Operation{Type: flow.ReplaceOperation, Flow: newICMPFlow(222), Key: strconv.Itoa(rand.Int())}
-	flowChan <- &flow.Operation{Type: flow.ReplaceOperation, Flow: newICMPFlow(444), Key: strconv.Itoa(rand.Int())}
+	flowChan <- newICMPFlow(222, "node1", "", "")
```
Comment (safchain) [inline gremlin/traversal/flow_test.go]: please do not change the test, please add epbf specific tests

File: flow/probes/ebpf.go (hunk 12 above)
Comment (nplanel) [reply flow/probes/ebpf.go]: old comment :
replace 2001 by 
p.flow_pool_size = cap(flowChanOperation)+1

File: flow/table.go (hunk 9 above)
Comment (andrekassis) [reply flow/table.go]: right now this library doesn't expose such a functionality, would you like me to override it? is it ok to do so (there are the licence issues or whatever). do you want us to open a PR?

File: flow/table.go (hunk 10 above)
Comment (andrekassis) [reply flow/table.go]: done

File: flow/table.go (hunk 11 above)
Comment (andrekassis) [reply flow/table.go]: done

File: flow/probes/ebpf.go (hunk 12 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: done

File: flow/table.go (hunk 13 above)
Comment (andrekassis) [reply flow/table.go]: done

File: flow/table.go (hunk 14 above)
Comment (andrekassis) [reply flow/table.go]: done

File: gremlin/traversal/flow_test.go (hunk 15 above)
Comment (andrekassis) [reply gremlin/traversal/flow_test.go]: done

File: flow/ebpf.go [hunk 16]
```diff
@@ -0,0 +1,290 @@
+package flow
```
Comment (nplanel) [inline flow/ebpf.go]: please add the license header

File: flow/probes/ebpf.go [hunk 17]
```diff
@@ -53,22 +50,16 @@ const (
...
-	expire       time.Duration
-	quit         chan bool
+	probeNodeTID   string
+	fd             int
+	flowTable      *flow.Table
+	module         *elf.Module
+	fmap           *elf.Map
+	expire         time.Duration
+	quit           chan bool
+	flow_pool_size int
```
Comment (nplanel) [inline flow/probes/ebpf.go]: keep the camelCase style flowPoolSize
Sorry, my fault, on my last comment ;)

File: flow/probes/ebpf.go [hunk 18]
```diff
@@ -80,298 +71,24 @@ type EBPFProbesHandler struct {
...
 	defer p.flowTable.Stop()
 
 	var startKTimeNs int64
 	var start time.Time
 
-	ebpfFlows := make(map[C.__u64]*EBPFFlow)
 	updateTicker := time.NewTicker(ebpfUpdate)
 	defer updateTicker.Stop()
 
+	p.flow_pool_size = 2*cap(flowChanOperation) + 1
```
Comment (nplanel) [inline flow/probes/ebpf.go]: why 2*cap ? the chan will block at cap so pool = cap+1 is enough

File: flow/ebpf.go [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/skydive-project/skydive/common"
+)
+
+// #cgo CFLAGS: -I../probe/ebpf
+// #include "flow.h"
+import "C"
```
Comment (nplanel) [inline flow/ebpf.go]: this is breaking windows cross compiling target
generate a dummy implentation of EBFFlow in no_ebpf.go with header // +build !linux

Comment (andrekassis) [reply flow/ebpf.go]: done

File: flow/probes/ebpf.go (hunk 17 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: done

File: flow/ebpf.go (hunk 16 above)
Comment (andrekassis) [reply flow/ebpf.go]: done

File: flow/probes/ebpf.go (hunk 18 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: this one is a bit tricky. It seems that go keeps local read/write buffers to use with channels so that when the channel blocks after cap entries and the reader performs a read operation the entire content of the channel is copied to its local buffer (for performance) and then reusing the pointers by the writer results in an overlap (I myself experienced this in one of the stress tests). therefore the pool size should be twice the size of the channel ( I agree that +1 here is redundant, can remove it if you'd like)

**Comment by andrekassis** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by andrekassis** [comment]
run skydive-functional-tests-backend-orientdb

File: flow/probes/ebpf.go (hunk 18 above)
Comment (nplanel) [reply flow/probes/ebpf.go]: golang channel use lockfree mechanism, but this should not be the double ... or I missing something ... do you have a bench test reproducing your issue (in case of cap+1) ?

Comment (nplanel) [reply flow/probes/ebpf.go]: I found your issue and 2*cap()+1 will not fix it

Imaging you can push 3 elements in the chan
len(chan) is 3
next is 3
 ... now the chan reader side read the 3 elements
len(chan) is 0
next is still 3

so you are out of sync between the chan pointer and pointer that could be already referenced in another element in the chan.

so you should remove next and have another channel to send back, the freed pointers; at start the allocated pointers should be injected in the _freed_ channel.

i := getFreePointer()   // <- freeEbpfFlows
write in &ebpfFlows[i]
flowChanOperation <- &ebpfFlows[i]

File: flow/flow.go [hunk 20]
```diff
@@ -309,28 +312,66 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
 	return layer.TransportFlow(), nil
 }
 
+//Hash flow with custom func, ignore ports in transport level for VXLAN and Geneve
+func Hash(f gopacket.Flow, hasher hash.Hash, isTransportUDP bool) {
+	src, dst := f.Endpoints()
+	if isTransportUDP {
+		srcPort, _ := strconv.ParseInt(src.String(), 10, 64)
+		dstPort, _ := strconv.ParseInt(dst.String(), 10, 64)
+		if srcPort == 4789 || srcPort == 6081 {
```
Comment (safchain) [inline flow/flow.go]: If you don't change the code line 368-369, you don't have to handle geneve nor vxlan here, IMHO

File: flow/flow.go [hunk 21]
```diff
@@ -309,28 +312,66 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
...
+			b := make([]byte, 8)
+			binary.BigEndian.PutUint64(b, uint64(linkID(p)))
+			hasher.Write(b)
 		}
 	}
 	if layer := p.NetworkLayer(); layer != nil {
-		uuid ^= layer.NetworkFlow().FastHash()
+		Hash(layer.NetworkFlow(), hasher, false)
+		b := make([]byte, 8)
+		binary.BigEndian.PutUint64(b, uint64(networkID(p)))
```
Comment (safchain) [inline flow/flow.go]: Why is this needed ?

File: flow/no_ebpf.go [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+package flow
+
+import (
+	"time"
+	"unsafe"
+)
+
+// EBPFFlow Wrapper type used for passing flows from probe to main agent routine
+type EBPFFlow struct {
+	start        time.Time
```
Comment (safchain) [inline flow/no_ebpf.go]: All the struct member are useless

File: flow/probes/ebpf.go [hunk 23]
```diff
@@ -80,298 +71,24 @@ type EBPFProbesHandler struct {
...
 	defer p.flowTable.Stop()
 
 	var startKTimeNs int64
 	var start time.Time
 
-	ebpfFlows := make(map[C.__u64]*EBPFFlow)
 	updateTicker := time.NewTicker(ebpfUpdate)
 	defer updateTicker.Stop()
 
+	p.flowPoolSize = 2*cap(flowChanOperation) + 1
```
Comment (safchain) [inline flow/probes/ebpf.go]: p := 2*cap....
not need to have it as struct member

File: flow/table.go (hunk 9 above)
Comment (safchain) [reply flow/table.go]: I need to check what the cost of this. Yes maybe opening a PR.

File: flow/table.go [hunk 24]
```diff
@@ -445,6 +458,42 @@ func (ft *Table) processFlowOP(op *Operation) {
...
+			ABRstStart: updateTCPFlagTime(fl.TCPMetric.ABRstStart, nfl.TCPMetric.ABRstStart),
+			BARstStart: updateTCPFlagTime(fl.TCPMetric.BARstStart, nfl.TCPMetric.BARstStart),
+		}
+	}
+	// TODO(safchain) remove this should be provided by the sender
+	// with a good time resolution
+	if fl.Metric.RTT == 0 && fl.Metric.ABPackets > 0 && fl.Metric.BAPackets > 0 {
+		fl.Metric.RTT = fl.Last - fl.Start
+	}
+
```
Comment (safchain) [inline flow/table.go]: nit: remove extra line

File: flow/probes/ebpf.go [hunk 25]
```diff
@@ -80,298 +71,24 @@ type EBPFProbesHandler struct {
...
-		Key:  key,
-	}
-}
-
 func (p *EBPFProbe) run() {
 	var info syscall.Sysinfo_t
 	syscall.Sysinfo(&info)
 
-	_, flowChanOperation := p.flowTable.Start()
+	_, flowChanOperation, _ := p.flowTable.Start()
```
Comment (safchain) [inline flow/probes/ebpf.go]: please change flowChanOperation to flowEBPFChan

File: flow/probes/ebpf.go (hunk 18 above)
Comment (eranra) [reply flow/probes/ebpf.go]: @nplanel @andrekassis is using circular buffer that the size of the buffer is twice the size of the channel he is writing to . You @nplanel are correct that next = 3 (in your example) but this is ok ... 3 is free to use ... and so is 4  and 5 (3*2 size) ... when new flows are written to 3,4,5 the reader will continue to read and this is ok because those are new flows ... when reader ends reading 3,4,5 next goes to 6 ... so circular goes to zero --- i.e. next = 0 and this is again ok, because we know that the reader already read 0,1 and 2 ... I think that 2 * capacity is ok (assuming the implementation of chan in golan has local buffer up to the size of the channel).

What you suggest will work ... I am sitting here with @andrekassis trying to review and I think this is fine. If we are missing something can you help by adding more elaborated example ... sorry for the long remark ... but we are trying to think and understand what is going on here.

File: flow/flow.go [hunk 26]
```diff
@@ -309,28 +312,66 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
 	return layer.TransportFlow(), nil
 }
 
+//Hash flow with custom func, ignore ports in transport level for VXLAN and Geneve
```
[the remark this answers is not in the document]
Comment (andrekassis) [reply flow/flow.go]: done

File: flow/flow.go [hunk 27]
```diff
@@ -309,28 +312,66 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
...
+func Hash(f gopacket.Flow, hasher hash.Hash, isTransportUDP bool) {
```
[the remark this answers is not in the document]
Comment (andrekassis) [reply flow/flow.go]: done

File: flow/flow.go (hunk 20 above)
Comment (andrekassis) [reply flow/flow.go]: maybe, but the whole reason I changed those lines was that the hash function that was being used wasn't working correctly, right?
one more thing though, if you take a look at the code of the hash function for this layer that was in use originally (which is in flows.go in gopacket's code) you won't see any special treatmen for those protocols, and yet the ports are always being hashed, which begs the question, how come this works?

File: flow/flow.go (hunk 21 above)
Comment (andrekassis) [reply flow/flow.go]: if you drop the last line, it would result in mismatching flows (those that have Geneve application layer, because we need the VNI in order to distinguish between them)
** you cant test it yourself -- make test will fail

File: flow/no_ebpf.go (hunk 22 above)
Comment (andrekassis) [reply flow/no_ebpf.go]: done

File: flow/probes/ebpf.go (hunk 23 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: +1 is redundant so I removed it;)
this is no longer a struct member as well.
As for your comment @nplanel -- I can indeed incorporate this change you are suggesting, but I still don't understand why it's needed. Eran also wrote a comment for u on this. Would be much appreciated if you could take a look, because both of us don't see why the current implementation is wrong.
+ adding a channel in the opposite direction is blocking as well and harms performance so would be smart to avoid this if the code is working as is, right?

File: flow/table.go (hunk 24 above)
Comment (andrekassis) [reply flow/table.go]: done

File: flow/probes/ebpf.go (hunk 25 above)
Comment (andrekassis) [reply flow/probes/ebpf.go]: done

File: flow/flow.go [hunk 28]
```diff
@@ -312,25 +312,30 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
...
-		uuid ^= layer.NetworkFlow().FastHash()
+		Hash(layer.NetworkFlow(), hasher, false)
+		b := make([]byte, 8)
+		binary.BigEndian.PutUint64(b, uint64(networkID(p)))
+		hasher.Write(b)
+
 	}
-	if tf, err := p.TransportFlow(); err == nil {
-		uuid ^= tf.FastHash()
+	if layer := p.TransportLayer(); layer != nil {
```
Comment (safchain) [inline flow/flow.go]: To follow my comment on the Hash function, previously TransportFlow was used and that were the geneve/vxlan thing was handled. I think you just need to use it instead of layer.TransportFlow so that you should be able to remove the geneve/vxlan thing from your hash function. You will see that the VNI is also used there.

File: flow/probes/ebpf.go [hunk 29]
```diff
@@ -49,17 +46,11 @@ import (
 import "C"
 
 const (
-	BPF_ANY    = 0
-	ebpfUpdate = 2 * time.Second
+	BPF_ANY       = 0
+	FLOW_TABLE_SZ = 500000
+	ebpfUpdate    = 10 * time.Second
```
Comment (andrekassis) [inline flow/probes/ebpf.go]: under large stress, i.e. hundreds of thousands of different flows or more, 2 seconds become just illogical (empirically) since the agent needs much longer than that to go over the entire table.
Changing the number to 10 make much more sense, it's much cheeper in CPU since the agent can rest (unlike 2 where the scans happen all the time and kill the system). 
When the number of connections is low we don't lose anything since we will finally read all the flows.
As for a large number of connections, it seems that the average HTTP server for example handles around 30K connections per second, so since we have a table of 500K that we scan every 10 s, the average number of flows per second we can track is 50K which is bigger.

File: flow/probes/ebpf.go [hunk 30]
```diff
@@ -394,50 +111,31 @@ func (p *EBPFProbe) run() {
...
+				found, err := p.module.LookupNextElement(p.fmap, unsafe.Pointer(&key), unsafe.Pointer(&nextKey), unsafe.Pointer(&kernFlows[nextAvailablePtr]))
 				if !found || err != nil {
+					p.module.DeleteElement(p.fmap, unsafe.Pointer(&key))
+					key = 0
 					break
 				}
-				key = nextKey
 				// delete every entry after we read the entry value
 				p.module.DeleteElement(p.fmap, unsafe.Pointer(&key))
+				key = nextKey
```
Comment (andrekassis) [inline flow/probes/ebpf.go]: we need to delete the previous key and not the new one, since deleting the new one will result in looking for a non existant key next time which will never fail but will always take us back to the beginning of the table (according to the bpf man) and therefore the loop will never end in some scenarios and also we will be starving flows

File: flow/probes/ebpf.go [hunk 31]
```diff
@@ -394,50 +111,31 @@ func (p *EBPFProbe) run() {
...
-			}
-
-			for k, v := range ebpfFlows {
-				if time.Now().Sub(v.last).Seconds() > p.expire.Seconds() {
-					delete(ebpfFlows, k)
+				flow.SetEBPFFlow(&ebpfFlows[nextAvailablePtr], us, last, unsafe.Pointer(&kernFlows[nextAvailablePtr]), startKTimeNs, p.probeNodeTID)
+				flowEBPFChan <- &ebpfFlows[nextAvailablePtr]
+				nextAvailablePtr = (nextAvailablePtr + 1) % flowPoolSize
+				flowsRead++
+				if flowsRead == FLOW_TABLE_SZ {
```
Comment (andrekassis) [inline flow/probes/ebpf.go]: don't scan more than 500K flows everytime (avoid infinite loop -- however, to avoid starvation, we now keep the index of the current flow and start from the same point next time)

**Comment by andrekassis** [comment]
run skydive-functional-tests-backend-elasticsearch

File: flow/flow.go (hunk 28 above)
Comment (andrekassis) [reply flow/flow.go]: you are correct again;)
done

**Comment by andrekassis** [comment]
skydive-functional-tests-backend-orientdb

**Comment by andrekassis** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by andrekassis** [comment]
run skydive-k8s-tests

File: flow/table.go [hunk 32]
```diff
@@ -120,11 +122,12 @@ func NewTable(updateHandler *Handler, expireHandler *Handler, nodeTID string, op
 		// convert seconds to milleseconds
 		appTimeout[strings.ToUpper(key)] = int64(1000 * config.GetConfig().GetInt("flow.application_timeout."+key))
 	}
-
+	LRU, _ := simplelru.NewLRU(500000, nil)
```
Comment (safchain) [inline flow/table.go]: This is now a strong limitation, while I'm fine with it I would add something in the config file to set it, default of 500k seems ok:

https://github.com/skydive-project/skydive/blob/master/etc/skydive.yml.default#L455

Comment (safchain) [reply flow/table.go]: I'm not sure about what advantage this implementation about LRU brings as internally it is a map, there is just a check that limit the amount of data and as there is a evictList it should take even more memory space. Probably limiting the amount of flows is better. Moreover there is no event when a flow is evicted so we can't mark it as timeout.

Comment (andrekassis) [reply flow/table.go]: Well, if you remember the previous implementation had no limit on the table's size which was causing it to inflate in size and kill the system. LRU implementation as we discussed enables you to ensure this doesn't happen and alliws you to keep the most recent flows as well. Also we can add a functionality to handle evictions if you'd like...

Comment (andrekassis) [reply flow/table.go]: Well, if you remember the previous implementation had no limit on the table's size which was causing it to inflate in size and kill the system. LRU implementation as we discussed enables you to ensure this doesn't happen and alliws you to keep the most recent flows as well. Also we can add a functionality to handle evictions if you'd like...

**Comment by safchain** [comment]
run skydive-k8s-tests
